lib/dynflow/action/with_polling_sub_plans.rb

```diff
@@ -0,0 +1,72 @@
+module Dynflow
+  module Action::WithPollingSubPlans
+
+    REFRESH_INTERVAL = 10
+    Poll = Algebrick.atom
+
+    def run(event = nil)
+      case event
+      when Poll
+        poll
+      else
+        super(event)
+      end
+    end
+
+    def poll
+      recalculate_counts
+      try_to_finish || suspend_and_ping
+    end
+
+    def initiate
+      ping suspended_action
+      super
+    end
+
+    def wait_for_sub_plans(sub_plans)
+      increase_counts(sub_plans.count, 0)
+      suspend
+    end
+
+    def resume
+      if sub_plans.all? { |sub_plan| sub_plan.error_in_plan? }
+        # We're starting over and need to reset the counts
+        %w(total failed pending success).each { |key| output.delete("#{key}_count".to_sym) }
+        initiate
+      else
+        if self.is_a?(::Dynflow::Actions::WithBulkSubPlans) && can_spawn_next_batch?
+          # Not everything was spawned
+          ping suspended_action
+          spawn_plans
+          suspend
+        else
+          poll
+        end
+      end
+    end
+
+    def notify_on_finish(_sub_plans)
+      suspend
+    end
+
+    def suspend_and_ping
+      suspend { |suspended_action| ping suspended_action }
+    end
+
+    def ping(suspended_action)
+      world.clock.ping suspended_action, REFRESH_INTERVAL, Poll
+    end
+
+    def recalculate_counts
+      total      = sub_plans.count
+      @sub_plans = nil # TODO:
```

Comment:
If they are dependent, having two commits in one PR, with the final set of changes would be the best.